DOI Content Negotiations.js

```diff
@@ -0,0 +1,205 @@
+
+	Copyright © 2016 Philipp Zumstein
+
+	This file is part of Zotero.
+
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero. If not, see <http://www.gnu.org/licenses/>.
+
+	***** END LICENSE BLOCK *****
+*/
+
+function detectSearch(items) {
+	if(!items) return false;
+
+	if(typeof items == 'string' || !items.length) items = [items];
+
+	for(var i=0, n=items.length; i<n; i++) {
+		if(!items[i]) continue;
+
+		if(items[i].DOI && ZU.cleanDOI(items[i].DOI)) return true;
+		if(typeof items[i] == 'string' && ZU.cleanDOI(items[i])) return true;
+	}
+
+	return false;
+}
+
+function filterQuery(items) {
+	if(!items) return [];
+
+	if(typeof items == 'string' || !items.length) items = [items];
+
+	//filter out invalid queries
+	var dois = [], doi;
+	for(var i=0, n=items.length; i<n; i++) {
+		if(items[i].DOI && (doi = ZU.cleanDOI(items[i].DOI)) ) {
+			dois.push(doi);
+		} else if(typeof items[i] == 'string' && (doi = ZU.cleanDOI(items[i])) ) {
+			dois.push(doi);
+		}
+	}
+	return dois;
+}
+
+function doSearch(items) {
+	var dois = filterQuery(items);
+	if(!dois.length) return;
+	processDOIs(dois);
+}
+
+var fixItemType = {
+	"journal-article": "article-journal", //http://doi.org/10.1136%2Fbmj.2.5914.335-a and crossref generally
+	"book-chapter": "chapter", //http://doi.org/10.1017/CBO9781316216453.176
+	"reference-entry": "chapter", //http://doi.org/10.1002/14651858.CD002966.pub3
+	"report-series": "report",
+	"misc": "article-journal" //DataCite
+}
+
+function fixJSON(text) {
+	try {
+		var item = JSON.parse(text);
+		//Z.debug("CSL_JSON" + text)
+		Z.debug(item)
+		if(fixItemType[item.type]) item.type = fixItemType[item.type];
+		//	Z.debug(item.container-title)
+		if (item.type == "report") {
+			Z.debug("here")
+			Z.debug(item['container-title'])
+			item.genre = item['container-title']
+		}
+		if (item.type != "article-journal" && item.type != "paper-conference") {
+			item.note = "DOI: " + item.DOI;
+		}
+		//Sometimes date is in created, not issued: 10.1017/CCOL0521858429.016
+		if (!item.issued.length && item.created) {
+			item.issued = item.created;
+		}
+		if(item.issued && item.issued.raw) item.issued.literal = item.issued.raw;
+		if(item.accessed && item.accessed.raw) item.accessed.literal = item.accessed.raw;
+		return JSON.stringify([item]);
+	} catch(e) {
+		return false;
+	}
+}
+
+function processDOIs(dois) {
+	var doi = dois.pop();
+	ZU.doGet('https://doi.org/' + encodeURIComponent(doi), function(text) {
```

Comment:
Is this encoding function the correct one to use here? (The `doi` is used as part of the URI and not as a URI component.)

Reply:
They list [certain specific characters](http://www.doi.org/doi_handbook/2_Numbering.html#2.5.2.3) that should be encoded, but in practice I suspect they're just doing a full decode and we can encode the whole thing. E.g., even for slashes, which they mostly show as unencoded, they mention that for a case like `/../` you might want to encode one of them

(And note that, without encoding, `10.1002/(SICI)1097-461X(1999)72:4<287::AID-QUA11>3.0.CO;2-Y` is a 400, even though it doesn't include any of the "mandatory" characters.)